input/system/selfhosted/syslog_handler.go

```diff
@@ -2,29 +2,73 @@ package selfhosted
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
 	"regexp"
 	"strconv"
 	"time"
 
 	"gopkg.in/mcuadros/go-syslog.v2"
 
+	"github.com/pganalyze/collector/config"
 	"github.com/pganalyze/collector/util"
 )
 
 var logLinePartsRegexp = regexp.MustCompile(`^\[(\d+)-(\d+)\] (.*)`)
 var logLineNumberPartsRegexp = regexp.MustCompile(`^\[(\d+)-(\d+)\]$`)
 
-func setupSyslogHandler(ctx context.Context, logSyslogServer string, out chan<- SelfHostedLogStreamItem, prefixedLogger *util.Logger) error {
+func setupSyslogHandler(ctx context.Context, config config.ServerConfig, out chan<- SelfHostedLogStreamItem, prefixedLogger *util.Logger) error {
+	logSyslogServer := config.LogSyslogServer
 	channel := make(syslog.LogPartsChannel)
 	handler := syslog.NewChannelHandler(channel)
 
 	server := syslog.NewServer()
 	server.SetFormat(syslog.RFC5424)
 	server.SetHandler(handler)
-	err := server.ListenTCP(logSyslogServer)
-	if err != nil {
-		return err
+
+	if config.LogSyslogServerCertPath != "" {
+		caPool := x509.NewCertPool()
+		if config.LogSyslogServerCAPath != "" {
+			ca, err := os.ReadFile(config.LogSyslogServerCAPath)
+			if err != nil {
+				return fmt.Errorf("failed to read a Certificate Authority file: %s", err)
+			}
+			if ok := caPool.AppendCertsFromPEM(ca); !ok {
+				return fmt.Errorf("failed to append a Certificate Authority")
+			}
+		}
+
+		cert, err := os.ReadFile(config.LogSyslogServerCertPath)
+		if err != nil {
+			return fmt.Errorf("failed to read a certificate file: %s", err)
+		}
+		key, err := os.ReadFile(config.LogSyslogServerKeyPath)
+		if err != nil {
+			return fmt.Errorf("failed to read a key file: %s", err)
+		}
+		tlsCert, err := tls.X509KeyPair(cert, key)
+		if err != nil {
+			return err
+		}
+
+		tlsConfig := tls.Config{
+			ClientAuth:   tls.VerifyClientCertIfGiven,
```

Comment:
You can check out what you can specify as `ClientAuth` in here: https://pkg.go.dev/crypto/tls#ClientAuthType

I'm debating what would be good to put here.

I don't think `RequireAndVerifyClientCert` is a good option. If we use `RequireAndVerifyClientCert` here, it requires the client to have a cert, too. With the case that Postgres -> rsyslog -> pganalyze collector syslog case, the client here is rsyslog and the collector syslog is a server. Setting up the certificate here ensures that at least the collector is trustworthy, and rsyslog will trust the collector to send the log (over TLS). I think it's an enough guarantee and I don't think the collector needs to require the client (rsyslog side) to be trustworthy.

`VerifyClientCertIfGiven` will give the option that the client can provide a certificate, but if they do, they need to have a valid one. I thought this would be a good option, but when I tested with Aiven, it didn't work well for some reason. Both `RequireAndVerifyClientCert` and `VerifyClientCertIfGiven` works well when I specify the client cert in Aiven side, but if I leave client certs empty, they both fail to deliver logs.

I think `RequireAnyClientCert` and `RequestClientCert` are both useless in our context so they are not good options. I actually tried `NoClientCert` too to see if it works, but it didn't work even when I provided the cert on Aiven side.

I think I'm misunderstanding some fundamental. I just realized that below is "ClientCAs" so specifying it with "ca_path" might be misreading. I'll look into this more tomorrow or next week.

Reply:
>Setting up the certificate here ensures that at least the collector is trustworthy, and rsyslog will trust the collector to send the log (over TLS).

I'm confused: isn't it rsyslog sending the logs to the collector?

> I think it's an enough guarantee and I don't think the collector needs to require the client (rsyslog side) to be trustworthy.

Well, it would be good to be able to trust the logs delivered to the collector, but I agree this is less critical.

In general, I think `VerifyClientCertIfGiven` seems like the right behavior for us. What was the error you got from the Aiven setup with this setting?

Reply:
> I'm confused: isn't it rsyslog sending the logs to the collector?

Yes that's correct, the following is a quick diagram. Over there, "client" and "server" are pretty much the same as how web server works with the browser. This ensures that the collector syslog server (remote server) is trustworthy and the rsyslog can send the log to such trustworthy server.

![image](https://github.com/pganalyze/collector/assets/911433/eb195e6d-870d-4597-882d-dc9ec0312e26)

However, go-syslog (and any other TLS connection) does support mutual TLS (mTLS), meaning that it requires that both the client and the server are trustworthy. In that case, the client also needs to provide a certificate like the following:

![image](https://github.com/pganalyze/collector/assets/911433/200aab22-ba0c-4e7f-aa64-230e6266ea84)

I think you already got this, and now I realize that my English was just not making sense. I meant "rsyslog will trust the collector (enough) to send the log (to the collector) over TLS". ... hope this makes more sense?

> What was the error you got from the Aiven setup with this setting?

I finally tracked this down, it was related to TlsPeerNameFunc, where essentially the go-syslog is expecting that the client is also providing a certificate. Providing nil to that fixed the issue.

Reply:
Ah, yep, I was misreading: I read that as "rsyslog will trust (the collector to send the log)" rather than "rsyslog will trust (the collector) to send the log", which is what you meant. Thanks for clarifying.